File: patternHunter/script/analyzeWord.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2014, Masahiro Yano 
# Licensed under the BSD licenses. 
# https://github.com/masayano 
#
# What is this?
#
#   [tab separated word count file"s": word, count]
#   -> [tab separated word slope/average file:
#               word,
#               slope of smoothed probability / average,
#               JSON word number history (not smoothed),
#               JSON word smoothed probability history]
#
import os
import sys
import numpy as np
import json
from datetime import datetime as dt

MIN_AVERAGE_PROBABILITY = 0.000001

def loadFile(\
        smoothedNumMatrix,\
        smoothedProbMatrix,\
        wordCounters,\
        totalWordNums,\
        fileName):
    wordCounter  = {}
    totalWordNum = 0
    for line in open(fileName, 'r'):
        parts = line.split('\t')
        word  = parts[0]
        count = int(parts[1])
        wordCounter[word] = count
        if not smoothedNumMatrix.has_key(word):
            smoothedNumMatrix [word] = []
            smoothedProbMatrix[word] = []
        totalWordNum += 1
    wordCounters .append(wordCounter)
    totalWordNums.append(float(totalWordNum))

def makeMatrix(\
        smoothedNumMatrix,\
        smoothedProbMatrix,\
        wordCounters,\
        totalWordNums):
    wordTypeNum = float(len(smoothedNumMatrix.keys()))
    for i in range(len(wordCounters)):
        wordCounter  = wordCounters [i]
        totalWordNum = totalWordNums[i]
        totalNumAfterSmoothed = totalWordNum + wordTypeNum
        for key in smoothedNumMatrix.keys():
            if wordCounter.has_key(key):
                smoothedNumber = wordCounter[key] + 1 # smoothing
            else:
                smoothedNumber = 1 # smoothing
            probability = smoothedNumber / totalNumAfterSmoothed
            smoothedNumMatrix [key].append(smoothedNumber)
            smoothedProbMatrix[key].append(probability)

def calcPValue(smoothedProbMatrix):
    pValues = {}
    results = {}
    for key, values_y in smoothedProbMatrix.items():
        values_x = range(len(values_y))
        x = np.array(values_x)
        y = np.array(values_y)
        z = np.polyfit(x, y, 1)
        average = np.average(y)
        results[key] = z[0] / average
    return results

def getDate():
    tdatetime = dt.now()
    return tdatetime.strftime('%Y_%m_%d_%H_%M')

def writeResults(\
        results,\
        smoothedNumMatrix,\
        smoothedProbMatrix):
    f = open('analysis/'+getDate()+'.log', 'w')
    for key, slope in sorted(results.items(), key=lambda x:x[1], reverse=True):
        probArray = [float("%.3g"%smoothedProb) for smoothedProb in smoothedProbMatrix[key]]
        averageProb = sum(probArray)/len(probArray)
        if averageProb > MIN_AVERAGE_PROBABILITY:
            numArray  = [smoothedNum-1 for smoothedNum  in smoothedNumMatrix[key]]
            f.write(key + '\t' + str(round(slope,4)) + '\t' + str(numArray) + '\t' + str(probArray) + '\n')
    f.close()

if __name__=='__main__':
    if not os.path.exists("analysis"):
        os.mkdir("analysis")
    smoothedNumMatrix  = {}
    smoothedProbMatrix = {}
    wordCounters  = []
    totalWordNums = []
    if len(sys.argv) >= 2:
        for fileName in sys.argv[1:]:
            print "[Load for analyzing]", fileName
            loadFile(\
                    smoothedNumMatrix,\
                    smoothedProbMatrix,\
                    wordCounters,\
                    totalWordNums,\
                    fileName)
        makeMatrix(\
                smoothedNumMatrix,\
                smoothedProbMatrix,\
                wordCounters,\
                totalWordNums)
        results = calcPValue(smoothedProbMatrix)
        writeResults(\
                results,\
                smoothedNumMatrix,\
                smoothedProbMatrix)
    else:
        print "Usage: python analyzeWord.py [word count file name(s)]"

File: patternHunter/src/analyzeWord.cpp
/*******************************************************
/
/  Copyright 2014, Masahiro Yano 
/  Licensed under the BSD licenses. 
/  https://github.com/masayano 
/
/  What is this?
/
/   [tab separated word count file"s": word, count]
/   -> [tab separated word slope/average file:
/               word,
/               slope of smoothed probability / average,
/               JSON word number history (not smoothed),
/               JSON word smoothed probability history]
/
/********************************************************/
#include <fstream>
#include <string>
#include <map>
#include <vector>
#include <boost/tokenizer.hpp>
#include <boost/lexical_cast.hpp>
#include <algorithm>
#include <cmath>
#include <ctime>
#include <sstream>
#include <iomanip>

const double MIN_AVERAGE_PROBABILITY = 1.0e-6;

typedef boost::char_separator<char> char_separator;
typedef boost::tokenizer<char_separator> tokenizer;
char_separator tabSep("\t");

const std::vector<std::string> splitString(const std::string& str) {
    std::vector<std::string> parts;
    tokenizer tokens(str, tabSep);
    for (const auto& part : tokens) {
        parts.push_back(part);
    }
    return parts;
}

void loadFile(
        std::map<std::string, std::vector<int> >&    smoothedNumMatrix,
        std::map<std::string, std::vector<double> >& smoothedProbMatrix,
        std::vector<std::map<std::string, int> >& wordCounters,
        std::vector<double>& totalWordNums,
        const std::string& fileName) {
    std::map<std::string, int> wordCounter;
    int totalWordNum = 0;
    std::ifstream ifs(fileName.c_str());
    std::string buf;
    while(ifs && std::getline(ifs, buf)) {
        const auto parts = splitString(buf);
        const auto& word  = parts[0];
        const auto& count = boost::lexical_cast<int>(parts[1]);
        wordCounter[word] = count;
        const auto& it = smoothedNumMatrix.find(word);
        if(it == std::end(smoothedNumMatrix)) {
            smoothedNumMatrix [word] = std::vector<int>();
            smoothedProbMatrix[word] = std::vector<double>();
        }
        totalWordNum += 1;
    }
    wordCounters .push_back(wordCounter);
    totalWordNums.push_back(static_cast<double>(totalWordNum));
}

void makeMatrix(
        std::map<std::string, std::vector<int> >&    smoothedNumMatrix,
        std::map<std::string, std::vector<double> >& smoothedProbMatrix,
        const std::vector<std::map<std::string, int> >& wordCounters,
        const std::vector<double>& totalWordNums) {
    const auto wordTypeNum = static_cast<double>(smoothedNumMatrix.size());
    for(int i = 0; i < wordCounters.size(); ++i) {
        const auto& wordCounter  = wordCounters [i];
        const auto& totalWordNum = totalWordNums[i];
        const auto& totalNumAfterSmoothed = totalWordNum + wordTypeNum;
        for(const auto& elem : smoothedNumMatrix) {
            const auto& key = elem.first;
            const auto& it  = wordCounter.find(key);
            int smoothedNumber;
            if(it != std::end(wordCounter)) {
                smoothedNumber = (*it).second + 1; // smoothing
            } else {
                smoothedNumber = 1; // smoothing
            }
            const auto probability = smoothedNumber / totalNumAfterSmoothed;
            smoothedNumMatrix [key].push_back(smoothedNumber);
            smoothedProbMatrix[key].push_back(probability);
        }
    }
}

const std::vector<double> range(int num) {
    std::vector<double> output;
    for(int i = 0; i < num; ++i) {
        output.push_back(i);
    }
    return output;
}

double calcSlope(
        const std::vector<double>& values_x,
        const std::vector<double>& values_y) {
    const auto length = static_cast<double>(values_x.size());
    const auto average_x = std::accumulate(std::begin(values_x), std::end(values_x), 0.0) / length;
    const auto average_y = std::accumulate(std::begin(values_y), std::end(values_y), 0.0) / length;
    double coVar = 0;
    double var_x = 0;
    for(int i = 0; i < length; ++i) {
        const auto xDiff = values_x[i] - average_x;
        const auto yDiff = values_y[i] - average_y;
        coVar += (xDiff * yDiff);
        var_x += (xDiff * xDiff);
    }
    return coVar / var_x;
}

const std::vector<std::pair<double, std::string> > calcSlope(
        const std::map<std::string, std::vector<double> >& smoothedProbMatrix) {
    std::vector<std::pair<double, std::string> > results;
    const auto length = (*std::begin(smoothedProbMatrix)).second.size();
    const auto values_x = range(length);
    for(const auto& elem : smoothedProbMatrix) {
        const auto& key      = elem.first;
        const auto& values_y = elem.second;
        const auto z = calcSlope(values_x, values_y);
        const auto average = std::accumulate(std::begin(values_y), std::end(values_y), 0.0) / length;
        results.push_back(std::make_pair(z / average, key));
    }
    std::sort(std::begin(results), std::end(results), std::greater<std::pair<double, std::string>>());
    return results;
}

const std::string getDate() {
    time_t now;
    time(&now);
    const auto date = localtime(&now);
    const auto year   = date->tm_year + 1900;
    const auto month  = date->tm_mon + 1;
    const auto day    = date->tm_mday;
    const auto hour   = date->tm_hour;
    const auto minute = date->tm_min;
    std::stringstream ss;
    ss << year << "_" << month << "_" << day << "_" << hour << "_" << minute;
    return ss.str();
}

const std::string makeStrProbArray(const std::vector<double>& probArray) {
    std::stringstream output;
    output << "[";
    for(int i = 0; i < probArray.size(); ++i) {
        output << std::setprecision(3) << probArray[i];
        if(i != probArray.size()-1) {
            output << ", ";
        }
    }
    output << "]";
    return output.str();
}

const std::string makeStrNumArray(const std::vector<int>& numArray) {
    std::stringstream output;
    output << "[";
    for(int i = 0; i < numArray.size(); ++i) {
        output << numArray[i] - 1;
        if(i != numArray.size()-1) {
            output << ", ";
        }
    }
    output << "]";
    return output.str();
}

void writeResults(
        const std::vector<std::pair<double, std::string> >& results,
        std::map<std::string, std::vector<int> >&    smoothedNumMatrix,
        std::map<std::string, std::vector<double> >& smoothedProbMatrix) {
    const auto fileName = "analysis/" + getDate() + ".log";
    std::ofstream ofs(fileName.c_str());
    for(const auto& elem : results) {
        const auto& slope = elem.first;
        const auto& key   = elem.second;
        const auto& probArray = smoothedProbMatrix[key];
        const auto strProbArray = makeStrProbArray(probArray);
        const auto averageProb = std::accumulate(std::begin(probArray), std::end(probArray), 0.0) / probArray.size();
        if(averageProb > MIN_AVERAGE_PROBABILITY) {
            const auto strNumArray = makeStrNumArray(smoothedNumMatrix[key]);
            ofs
                    << key
                    << "\t"
                    << std::setprecision(3) << slope
                    << "\t"
                    << strNumArray
                    << "\t"
                    << strProbArray
                    << std::endl;
        }
    }
}

int main(int argc, char** argv) {
    std::map<std::string, std::vector<int> >    smoothedNumMatrix;
    std::map<std::string, std::vector<double> > smoothedProbMatrix;
    std::vector<std::map<std::string, int> >    wordCounters;
    std::vector<double>                         totalWordNums;
    if(argc >= 2) {
        for(int i = 1; i < argc; ++i) {
            const auto fileName = argv[i];
            std::cout << "[Load for analyzing] " << fileName << std::endl;
            loadFile(
                    smoothedNumMatrix,
                    smoothedProbMatrix,
                    wordCounters,
                    totalWordNums,
                    fileName);
        }
        makeMatrix(
                smoothedNumMatrix,
                smoothedProbMatrix,
                wordCounters,
                totalWordNums);
        const auto results = calcSlope(smoothedProbMatrix);
        writeResults(
                results,
                smoothedNumMatrix,
                smoothedProbMatrix);
    } else {
        std::cout << "Usage: python analyzeWord.py [word count file name(s)]" << std::endl;
    }
    return 0;
}
